README.md
# AutoSpotting #
<!-- markdownlint-disable MD003 MD026 MD033 -->

<img src="logo.png" width="150" align="right">

[![BuildStatus](https://travis-ci.org/AutoSpotting/AutoSpotting.svg?branch=master)](https://travis-ci.org/AutoSpotting/AutoSpotting)
[![GoReportCard](https://goreportcard.com/badge/github.com/AutoSpotting/AutoSpotting)](https://goreportcard.com/report/github.com/AutoSpotting/AutoSpotting)
[![CoverageStatus](https://coveralls.io/repos/github/AutoSpotting/AutoSpotting/badge.svg?branch=master)](https://coveralls.io/github/AutoSpotting/AutoSpotting?branch=master)
[![CodeClimate](https://codeclimate.com/github/AutoSpotting/AutoSpotting/badges/gpa.svg)](https://codeclimate.com/github/AutoSpotting/AutoSpotting)
[![IssueCount](https://codeclimate.com/github/AutoSpotting/AutoSpotting/badges/issue_count.svg)](https://codeclimate.com/github/AutoSpotting/AutoSpotting)
[![ChatOnGitter](https://badges.gitter.im/AutoSpotting/AutoSpotting.svg)](https://gitter.im/cristim/autospotting)
[![Open Source Helpers](https://www.codetriage.com/AutoSpotting/AutoSpotting/badges/users.svg)](https://www.codetriage.com/AutoSpotting/AutoSpotting)
[![Patreon](https://img.shields.io/badge/patreon-donate-yellow.svg)](https://www.patreon.com/cristim/overview)

AutoSpotting is the leading open source spot market automation tool, optimized towards quick/easy/frictionless adoption of the EC2 spot market at any scale.

It takes over existing long-running ASGs with minimal configuration changes(usually just tagging them, but even that can be avoided), yielding the usual 70%-90% cost savings on compute but in a better integrated, more cost effective and easier to adopt way than the commercial SaaS alternatives in this space and for some mass adoption use cases it's even a better fit than the current native AWS offerings.

## How does it work? ##

Once installed and enabled (usually by tagging) on existing on-demand AutoScaling groups,
AutoSpotting gradually replaces their on-demand instances with cheaper spot instances
that are at least as large and identically configured to
the group's members, without changing the group launch configuration in any way. For
your peace of mind, you can also keep running a configurable number of on-demand
instances given as percentage or absolute number and it automatically fails over to on-demand in case of spot instance terminations.

Going forward, as well as on any new ASGs that match the expected tags, any new on-demand instances above the amount configured to be keept will be immediately replaced with spot clones within seconds of being launched. If this fails due to insufficient spot capacity, the usual cron events will replace them later once spot capacity becomes available again. When launching Spot instances, the compatible instance types are attempted in increasing order of their price, until one is successfully launched.

This process can partly be seen in action below, you can click to expand the animation:

![Workflow](https://autospotting.org/img/autospotting.gif)

Additionally, it implements some complex logic aware of spot and on demand prices, including
for different spot products and configurable discounts for reserved instances or
large volume customers. It also considers the specs of all instance types and
automatically places bids to instance types and prices chosen based on flexible
configuration set globally or overridden at the group level using additional
tags, but these overrides are often not needed.

A single installation can handle all enabled groups fronm an AWS account in parallel across all
available AWS regions, but it can be restricted to fewer regions if desired.

Your groups will then monitor and use these spot instances just like they would
do with your on-demand instances. They will automatically join your load
balancer and start receiving traffic once passing the health checks, and the
traffic would automatically be drained on termination.

## What savings can I expect? ##

The savings it generates are often in the 60-80% range, depending on region and instance type, but sometimes even up to
90% of on-demand prices, like you can see in the graph below.

![Savings](https://autospotting.org/img/savings.png)

## What's under the hood? ##

The entire logic described above is implemented in a set of Lambda functions deployed
using CloudFormation or Terraform stacks that can be installed and configured in
just a few minutes.

The stack assigns them the minimal set of IAM permissions required for
them to work and requires no admin-like cross-account permissions. The entire code
base can be audited to see how these permissions are being used and even locked
down further if your audit discovers any issues. **This is not a SaaS**, there's
no component that calls home and reveals any details about your infrastructure.

The main Lambda function is written in the Go programming language and the code is
compiled as a static binary compressed and uploaded to S3. For evaluation or
debugging purposes, the same binary can run out of the box locally on Linux
machines or as a Docker container on Windows or macOS. Some people even run
these containers on their existing Kubernetes clusters assuming the other
resources provided by the stack are implemented in another way on Kubernetes.

The stack also consists of a few CloudWatch event triggers, that run the Lambda
function periodically and whenever it needs to take action against the enabled groups.
Between runs your group is entirely managed by AutoScaling (including any scaling policies
you may have) and load balancer health checks, that can trigger instance
launches or replacements using the original on-demand launch configuration.

Read [here](TECHNICAL_DETAILS.md) for more information and implementation
details.

## FAQs ##

Frequently asked questions about the project are answered in the
[FAQ](https://autospotting.org/faq/index.html), *please read this first before
asking for support*.

If you have additional questions not covered there, they can be easily added to
the
[source](https://github.com/AutoSpotting/autospotting.org/blob/master/content/faq.md)
of the FAQ by editing in the browser and creating a pull request, and we'll
answer them while reviewing the pull request.

## Getting Started ##

Just like in the above animation, it's as easy as launching a CloudFormation (or
[Terraform](https://github.com/AutoSpotting/terraform-aws-autospotting))
stack and setting the (configurable) `spot-enabled` tag on the AutoScaling
groups where you want it enabled to `true`.

All the required infrastructure and configuration will be created automatically,
so you can get started as fast as possible.

For more detailed information you can read this [document](START.md)

[![Launch](https://s3.amazonaws.com/cloudformation-examples/cloudformation-launch-stack.png)](https://autospotting.org)

### Note ###

- the binaries launched by this stack are distributed under a proprietary
  [license](BINARY_LICENSE), and are **free to use for evaluation, up to $1000
  monthly savings**. Once you reach this limit you'll need to either switch to the
  inexpensive supported binaries (designed to cost a small fraction of around
  1% of your total savings for supporting further development), or you can build
  your own binaries based on the open source code and run it for free.

## Support ##

Community support is available on the
[gitter](https://gitter.im/cristim/autospotting) chat room, where the main
authors and other users are likely to help you solve issues.

Note: This is offered on a best effort basis and under certain conditions, such
as using the latest version of the evaluation binaries.

If you need more comprehensive support you will need to purchase a support plan.

## Contributing ##

Unlike multiple commercial products in this space that cost a lot of money and
attempt to lock you in, this project is fully open source and developed in the
open by a vibrant community of dozens of contributors.

We urge you to support us on [Github Sponsors](https://github.com/sponsors/cristim)
if this software helps you save any significant amount of money, this will greatly help
further development.

Financial sponsorship is voluntary, it's also fine if you just try it out and give
[feedback](https://gitter.im/cristim/autospotting), report issues, improve the
documentation, write some code or assign a developer to work on it, or even just
spread the word among your peers who might be interested in it. Any sort of
support would be greatly appreciated and would make a huge difference
to the project.

Note: Non-trivial code should be submitted according to the contribution
[guidelines](CONTRIBUTING.md).

### Proprietary binaries ###

The source code is and will always be open source, so you can build and run
it yourself, see how it works and even enhance it if you want.

But if you want to conveniently get started or update within minutes without
setting up and maintaining a build environment or any additional infrastructure,
we have pre-built evaluation binaries that will save you significant
amounts of time and effort.

These can be used for evaluation purposes as long as the generated monthly
savings are less than $1000. Once you reach this level you will need to either
purchase an inexpensive stable build that doesn't have this limitation, and also
comes with a support plan, or you can build AutoSpotting from source code.

The support license costs vary by group, region and AWS account coverage and can
also be paid through [Patreon](https://www.patreon.com/cristim/overview).

Individuals and companies supporting the development of the open source code get
free of charge support and stable build access for a year since their latest
contribution to the project.

#### Note: ####

- even though these evaluation builds are usually stable enough, they may
  not have been thoroughly tested yet and come with best effort community
  support.
- the docker images available on DockerHub are also distributed under the
  same binary license and the costs are the same.

### Stable builds ###

Carefully tested builds suitable for Enterprise use will be communicated
to [Patreon](https://www.patreon.com/cristim/overview) backers as soon as
they join.

They come with support from the author, who will do his best to help you
successfully run AutoSpotting on your environment so you can get the most out of
it. The feature requests and issues will also be prioritized based on the
Patreon tier.

Please get in touch on [gitter](https://gitter.im/cristim) if you have any
questions about these stable builds.

## Compiling and Installing ##

It is recommended to use the evaluation or stable binaries, which are easy to
install, support further development of the software and allow you to get
support.

But if you have some special needs that require some customizations or you don't
want to rely on the author's infrastructure or contribute anything for longer
term use of the software, you can always build and run your customized binaries
that you maintain on your own, just keep in mind that those won't be supported
in any way.

More details are available [here](CUSTOM_BUILDS.md)

## Users ##

Autospotting is already used by hundreds of individuals and organizations around
the world and as per internal AWS information a significant(but undisclosed) 
percentage of all currently running spot instances were launched using it.

## License ##

This software is distributed under the terms of the OSL-3.0 [license](LICENSE).

The official binaries are licensed under this proprietary
[license](BINARY_LICENSE).


CHANGELOG.md
# ChangeLog

<!-- markdownlint-disable MD024 -->

Since 2019 the release notes, potential breaking changes and mitigation
procedures have only been communicated via private channels to the Patreon
subscribers when releasing a new stable version.

Sign up on [Patreon](https://www.patreon.com/cristim) in order to receive these
release notifications, as well as to get access to these supported stable
versions. In the future certain premium features will only be available in the
stable builds.

## 19 September 2018, build 942

Yet another really beefy entry, but most of this was also captured in much more
detail in this blog
[post](https://mcristi.wordpress.com/2018/07/14/new-autospotting-version/).

Many thanks to HERE Technologies for supporting a lot of the development effort,
to the Patreon supporters and to everyone who contributed code to this release.

### New features since the last update

- The method by which AutoSpotting terminates existing Auto Scaling Group
  instances has changed. By default, AutoSpotting now uses the EC2
  `TerminateInstanceInAutoScalingGroup` API. This API call ensures that any
  Termination Lifecycle Hooks that might be configured on the Group are
  respected, which was not the case in previous AutoSpotting versions. Users who
  depend upon the legacy behavior, which was to detach the instance from the
  Auto Scaling Group and terminate it, can set `instance_termination_method` to
  `detach` in their deployment configurations.
- Better handling of out of capacity situations.
- Better handling of VPC, DefaultVPC and EC2 Classic security groups.
- Support running in opt-out mode.
- Tagging launched spot instances with `launched-by-autospotting=true`.
- Obey the scale-in protection and termination protection for on-demand
  instances that previously were being replaced.
- Terraform module in the Terraform Registry.
- Documentation for running as a Kubernetes cron job instead of Lambda.
- Instance type updates and regional expansions. Note: some recent instance
  types are still not supported yet due to missing upstream pricing information.

### Under the hood changes

- AutoSpotting now launches spot instances using the `RunInstances` API call,
  which allows us to simplify the logic considerably and fix a number of bugs.
- The spot bidding engine was heavily refactored, using less memory and being
  much more scalable on large installations.
- Fix compilation of macOS.
- Smaller binaries.
- Build AutoSpotting using Go 1.11

## 24 January 2018, build 622

A lot of time passed since the previous Changelog update, so this is a really
beefy entry, we should definitely do this more often.

There were a lot of contributions from many individuals, thank you all for
helping improve this project!

Special thanks once again to @xlr-8 who contributed a lot throughout the
previous year, and is now a very active co-maintainer of the project.

### New features since the last update

- Consider GPUs in the instance compatbility checks
- Instance type updates and regional expansions
- Aggressive bidding strategy based on the current spot price, by @kartik894
- Prevent termination of last instance, by @universam1
- Support volume discounts (think Reserved Instance), by @masneyb
- Instance type whitelisting, by @raravena80
- Instance type blacklisting, by @binarylogic
- Consider the EBS pricing surcharge, by @vecchp
- Deploy using Terraform, by @xlr-8

### Under the hood changes

- Significant refactorings, especially done by @xlr-8
- We're now using the native Lambda Go runtime, again thanks to @xlr-8
- Scalability fixes: concurrency by @thebigjc, API pagination by @ahaverbuch,
  support handling more than 50 AutoScaling groups concurrently, by @chaner
- Much more unit test coverage, thanks to @xlr-8 and @artemnikitin
- Buildsystem changes, we now also have automated lint and vet checks
- Dependency vendoring, by @xlr-8
- Lots of bugfixes, special thanks to @xlr-8
- Countless documentation updates by too many individuals to mention here

### Breaking changes

- The CloudFormation stack needs to be updated after build 633, because of the
  change to the Go Lambda runtime
- Some of the AutoScaling group tags used to override the global configuration
  were renamed to be more consistent, please refer to the current state of the
  documentation.

~Cristian

## 29 December 2016, build 158

I forgot to update this in a while, so this is a quite big changelog entry.

I got the first major code contributions by other developers, so from now on the
changelog entries will be split by author. It may also have a header like this
one in which it will contain a short summary or a message from the author.

Special thanks to Hugo Rosnet, who contributed a lot of code that implemented a
number of major features, helped me with multiple code reviews and kept me
motivated enough to constantly work on this project.

Also thanks to Jay Wineinger who contributed a non-trivial piece of code, and to
the other folks who contributed documentation, raised or discussed various
Github issues.

~Cristian

### Changes by author

#### @cristim

- Big code refactoring to make the code more maintainable and testable.
- Buildsystem improvements (and regressions, since fixed).
- Updated regional and instance type coverage, thanks to ec2instances.info.
- Support restricting the execution to a given set of regions.
- Expose all configuration options also as CloudFormation stack parameters.
- Documentation updates and improvements.
- Random small cleanups.

#### @xlr-8

- Update Lambda function's IAM permissions.
- The algorithm now supports keeping on-demand instances in each AutoScaling
  group.
- The algorithm is now configurable using tags set on the group and based on
  flags when executing it locally as a CLI tool.
- Significant test coverage increases.
- Significant clean-up and refactoring of the core algorithm.
- Documentation improvements.

#### @jwineinger


core/config.go
// Copyright (c) 2016-2019 Cristian Măgherușan-Stanciu
// Licensed under the Open Software License version 3.0

package autospotting

import (
	"fmt"
	"io"
	"log"
	"os"
	"time"

	"github.com/aws/aws-sdk-go/aws/endpoints"
	ec2instancesinfo "github.com/cristim/ec2-instances-info"
	"github.com/namsral/flag"
)

const (
	// AutoScalingTerminationMethod uses the TerminateInstanceInAutoScalingGroup
	// API method to terminate instances.  This method is recommended because it
	// will require termination Lifecycle Hooks that have been configured on the
	// Auto Scaling Group to be invoked before terminating the instance.  It's
	// also safe even if there are no such hooks configured.
	AutoScalingTerminationMethod = "autoscaling"

	// DetachTerminationMethod detaches the instance from the Auto Scaling Group
	// and then terminates it.  This method exists for historical reasons and is
	// no longer recommended.
	DetachTerminationMethod = "detach"

	// TerminateTerminationNotificationAction terminate the spot instance, which will be terminated
	// by AWS in 2 minutes, without reducing the ASG capacity, so that a new instance will
	// be launched. LifeCycle Hooks are triggered.
	TerminateTerminationNotificationAction = "terminate"

	// DetachTerminationNotificationAction detach the spot instance, which will be terminated
	// by AWS in 2 minutes, without reducing the ASG capacity, so that a new instance will
	// be launched. LifeCycle Hooks are not triggered.
	DetachTerminationNotificationAction = "detach"

	// AutoTerminationNotificationAction if ASG has a LifeCycleHook with LifecycleTransition = EC2_INSTANCE_TERMINATING
	// terminate the spot instance (as TerminateTerminationNotificationAction), if not detach it.
	AutoTerminationNotificationAction = "auto"

	// DefaultCronSchedule is the default value for the execution schedule in
	// simplified Cron-style definition the cron format only accepts the hour and
	// day of week fields, for example "9-18 1-5" would define the working week
	// hours. AutoSpotting will only run inside this time interval. The action can
	// also be reverted using the CronScheduleState parameter, so in order to run
	// outside this interval set the CronScheduleState to "off" either globally or
	// on a per-group override.
	DefaultCronSchedule = "* *"
)

// Config extends the AutoScalingConfig struct and in addition contains a
// number of global flags.
type Config struct {
	AutoScalingConfig

	// Static data fetched from ec2instances.info
	InstanceData *ec2instancesinfo.InstanceData

	// Logging
	LogFile io.Writer
	LogFlag int

	// The regions where it should be running, given as a single CSV-string
	Regions string

	// The region where the Lambda function is deployed
	MainRegion string

	// This is only here for tests, where we want to be able to somehow mock
	// time.Sleep without actually sleeping. While testing it defaults to 0 (which won't sleep at all), in
	// real-world usage it's expected to be set to 1
	SleepMultiplier time.Duration

	// Filter on ASG tags
	// for example: spot-enabled=true,environment=dev,team=interactive
	FilterByTags string
	// Controls how are the tags used to filter the groups.
	// Available options: 'opt-in' and 'opt-out', default: 'opt-in'
	TagFilteringMode string

	// The AutoSpotting version
	Version string

	// The license of this AutoSpotting build
	LicenseType string

	// Controls whether AutoSpotting patches Elastic Beanstalk UserData scripts to use
	// the instance role when calling CloudFormation helpers instead of the standard CloudFormation
	// authentication method
	PatchBeanstalkUserdata string

	// Lambda to use for Managing ASG
	LambdaManageASG string

	// JSON file containing event data used for locally simulating execution from Lambda.
	EventFile string

	// Final Recap String Array to show actions taken by ScheduleRun on ASGs
	FinalRecap map[string][]string
}

// ParseConfig loads configuration from command line flags, environments variables, and config files.
func ParseConfig(conf *Config) {

	// The use of FlagSet allows us to parse config multiple times, which is useful for unit tests.
	flagSet := flag.NewFlagSet("AutoSpotting", flag.ExitOnError)

	var region string

	if r := os.Getenv("AWS_REGION"); r != "" {
		region = r
	} else {
		region = endpoints.UsEast1RegionID
	}

	conf.LogFile = os.Stdout
	conf.LogFlag = log.Ldate | log.Ltime | log.Lshortfile
	conf.MainRegion = region
	conf.SleepMultiplier = 1
	conf.LambdaManageASG = os.Getenv("LAMBDA_MANAGE_ASG")

	flagSet.StringVar(&conf.AllowedInstanceTypes, "allowed_instance_types", "",
		"\n\tIf specified, the spot instances will be searched only among these types.\n\tIf missing, any instance type is allowed.\n"+
			"\tAccepts a list of comma or whitespace separated instance types (supports globs).\n"+
			"\tExample: ./AutoSpotting -allowed_instance_types 'c5.*,c4.xlarge'\n")

	flagSet.StringVar(&conf.BiddingPolicy, "bidding_policy", DefaultBiddingPolicy,
		"\n\tPolicy choice for spot bid. If set to 'normal', we bid at the on-demand price(times the multiplier).\n"+
			"\tIf set to 'aggressive', we bid at a percentage value above the spot price \n"+
			"\tconfigurable using the spot_price_buffer_percentage.\n")

	flagSet.StringVar(&conf.DisallowedInstanceTypes, "disallowed_instance_types", "",
		"\n\tIf specified, the spot instances will _never_ be of these types.\n"+
			"\tAccepts a list of comma or whitespace separated instance types (supports globs).\n"+
			"\tExample: ./AutoSpotting -disallowed_instance_types 't2.*,c4.xlarge'\n")

	flagSet.StringVar(&conf.InstanceTerminationMethod, "instance_termination_method", DefaultInstanceTerminationMethod,
		"\n\tInstance termination method.  Must be one of '"+DefaultInstanceTerminationMethod+"' (default),\n"+
			"\t or 'detach' (compatibility mode, not recommended)\n")

	flagSet.StringVar(&conf.TerminationNotificationAction, "termination_notification_action", DefaultTerminationNotificationAction,
		"\n\tTermination Notification Action.\n"+
			"\tValid choices:\n"+
			"\t'"+DefaultTerminationNotificationAction+
			"' (terminate if lifecyclehook else detach) | 'terminate' (lifecyclehook triggered)"+
			" | 'detach' (lifecyclehook not triggered)\n")

	flagSet.Int64Var(&conf.MinOnDemandNumber, "min_on_demand_number", DefaultMinOnDemandValue,
		"\n\tNumber of on-demand nodes to be kept running in each of the groups.\n\t"+
			"Can be overridden on a per-group basis using the tag "+OnDemandNumberLong+".\n")

	flagSet.Float64Var(&conf.MinOnDemandPercentage, "min_on_demand_percentage", 0.0,
		"\n\tPercentage of the total number of instances in each group to be kept on-demand\n\t"+
			"Can be overridden on a per-group basis using the tag "+OnDemandPercentageTag+
			"\n\tIt is ignored if min_on_demand_number is also set.\n")

	flagSet.Float64Var(&conf.OnDemandPriceMultiplier, "on_demand_price_multiplier", DefaultOnDemandPriceMultiplier,
		"\n\tMultiplier for the on-demand price. Numbers less than 1.0 are useful for volume discounts.\n"+
			"The tag "+OnDemandPriceMultiplierTag+" can be used to override this on a group level.\n"+
			"\tExample: ./AutoSpotting -on_demand_price_multiplier 0.6 will have the on-demand price "+
			"considered at 60% of the actual value.\n")

	flagSet.StringVar(&conf.Regions, "regions", "",
		"\n\tRegions where it should be activated (separated by comma or whitespace, also supports globs).\n"+
			"\tBy default it runs on all regions.\n"+
			"\tExample: ./AutoSpotting -regions 'eu-*,us-east-1'\n")

	flagSet.Float64Var(&conf.SpotPriceBufferPercentage, "spot_price_buffer_percentage", DefaultSpotPriceBufferPercentage,
		"\n\tBid a given percentage above the current spot price.\n\tProtects the group from running spot"+
			"instances that got significantly more expensive than when they were initially launched\n"+
			"\tThe tag "+SpotPriceBufferPercentageTag+" can be used to override this on a group level.\n"+
			"\tIf the bid exceeds the on-demand price, we place a bid at on-demand price itself.\n")

	flagSet.StringVar(&conf.SpotProductDescription, "spot_product_description", DefaultSpotProductDescription,
		"\n\tThe Spot Product to use when looking up spot price history in the market.\n"+
			"\tValid choices: Linux/UNIX | SUSE Linux | Windows | Linux/UNIX (Amazon VPC) | \n"+
			"\tSUSE Linux (Amazon VPC) | Windows (Amazon VPC) | Red Hat Enterprise Linux\n\tDefault value: "+DefaultSpotProductDescription+"\n")

	flagSet.Float64Var(&conf.SpotProductPremium, "spot_product_premium", DefaultSpotProductPremium,
		"\n\tThe Product Premium to apply to the on demand price to improve spot selection and savings calculations\n"+
			"\twhen using a premium instance type such as RHEL.")

	flagSet.StringVar(&conf.TagFilteringMode, "tag_filtering_mode", "opt-in", "\n\tControls the behavior of the tag_filters option.\n"+
		"\tValid choices: opt-in | opt-out\n\tDefault value: 'opt-in'\n\tExample: ./AutoSpotting --tag_filtering_mode opt-out\n")

	flagSet.StringVar(&conf.FilterByTags, "tag_filters", "", "\n\tSet of tags to filter the ASGs on.\n"+
		"\tDefault if no value is set will be the equivalent of -tag_filters 'spot-enabled=true'\n"+
		"\tIn case the tag_filtering_mode is set to opt-out, it defaults to 'spot-enabled=false'\n"+
		"\tExample: ./AutoSpotting --tag_filters 'spot-enabled=true,Environment=dev,Team=vision'\n")

	flagSet.StringVar(&conf.CronSchedule, "cron_schedule", DefaultCronSchedule, "\n\tCron-like schedule in which to"+
		"\tperform(or not) spot replacement actions. Format: hour day-of-week\n"+
		"\tExample: ./AutoSpotting --cron_schedule '9-18 1-5' # workdays during the office hours \n")

	flagSet.StringVar(&conf.CronTimezone, "cron_timezone", "UTC", "\n\tTimezone to"+
		"\tperform(or not) spot replacement actions. Format: timezone\n"+
		"\tExample: ./AutoSpotting --cron_timezone 'Europe/London' \n")

	flagSet.StringVar(&conf.CronScheduleState, "cron_schedule_state", "on", "\n\tControls whether to take actions "+
		"inside or outside the schedule defined by cron_schedule. Allowed values: on|off\n"+
		"\tExample: ./AutoSpotting --cron_schedule_state='off' --cron_schedule '9-18 1-5'  # would only take action outside the defined schedule\n")

	flagSet.StringVar(&conf.LicenseType, "license", "evaluation", "\n\tControls the terms under which you use AutoSpotting"+
		"Allowed values: evaluation|I_am_supporting_it_on_Patreon|I_contributed_to_development_within_the_last_year|I_built_it_from_source_code\n"+
		"\tExample: ./AutoSpotting --license evaluation\n")

	flagSet.StringVar(&conf.EventFile, "event_file", "", "\n\tJSON file containing event data, "+
		"used for locally simulating execution from Lambda. AutoSpotting now expects to be "+
		"triggered by events and won't do anything if no event is passed either as result of "+
		"AWS instance state notifications or simulated manually using this flag.\n")

	flagSet.StringVar(&conf.PatchBeanstalkUserdata, "patch_beanstalk_userdata", "", "\n\tControls whether AutoSpotting patches Elastic Beanstalk UserData scripts to use the instance role when calling CloudFormation helpers instead of the standard CloudFormation authentication method\n"+
		"\tExample: ./AutoSpotting --patch_beanstalk_userdata true\n")

	flagSet.StringVar(&conf.LambdaManageASG, "lambda_manage_asg", "", "\n\tThe name of the Lambda function used to manage the ASG maximum group capacity. This needs to exist in the same region as the main AutoSpotting Lambda function"+
		"\tExample: ./AutoSpotting --lambda_manage_asg AutoSpotting-LambdaManageASG-01234567890ABC\n")

	printVersion := flagSet.Bool("version", false, "Print version number and exit.\n")

	if err := flagSet.Parse(os.Args[1:]); err != nil {
		fmt.Printf("Error parsing config: %s\n", err.Error())
	}

	if *printVersion {
		fmt.Println("AutoSpotting build:", conf.Version)
		os.Exit(0)
	}

	data, err := ec2instancesinfo.Data()
	if err != nil {
		log.Fatal(err.Error())
	}
	conf.InstanceData = data

	conf.FinalRecap = make(map[string][]string)
}
